### ISSUE TITLE
Simple Java + JML with System.out.println causes multiple warnings from internal JML specifications of the JDK

### ISSUE THREAD

**Issue by leavens** [OpenJML/OpenJML#789]
When I try to run ESC on the attached CircleOperations.java file I get the following output:
`$ esc CircleOperations.java
CircleOperations.java:9: verify: The prover cannot establish an assertion (InvariantLeaveCaller: /usr/local/openjml/specs/java/io/PrintStream.jml:42:) in method main: (Caller: CircleOperations.main(java.lang.String[]), Callee: java.io.PrintStream.println(int))
        System.out.println(cops.id(79));
                          ^
/usr/local/openjml/specs/java/io/PrintStream.jml:42: verify: Associated declaration: CircleOperations.java:9:
    //-RAC@ public invariant outputText != null && \invariant_for(outputText);
                   ^
CircleOperations.java:9: verify: The prover cannot establish an assertion (InvariantLeaveCaller: /usr/local/openjml/specs/java/io/PrintStream.jml:35:) in method main: (Caller: CircleOperations.main(java.lang.String[]), Callee: java.io.PrintStream.println(int))
        System.out.println(cops.id(79));
                          ^
/usr/local/openjml/specs/java/io/PrintStream.jml:35: verify: Associated declaration: CircleOperations.java:9:
    //-RAC@ public static invariant \invariant_for(eol) && eol.length() > 0;
                          ^
4 verification failures
$ esc --version
openjml 0.17.0-alpha-15
$ uname -a
Linux NEW-SURFACE 5.10.16.3-microsoft-standard-WSL2 #1 SMP Fri Apr 2 22:23:49 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`
[CircleOperations.zip](https://github.com/OpenJML/OpenJML/files/10189794/CircleOperations.zip)

**Comment by marcel139** [comment]
Hi Gary,

I am facing the same problem. Did you have any success on dealing with this problem?

Regards,
Marcel

**Comment by leavens** [comment]
Hi Marcel,

No, nothing has been done about that yet.


        Regards,

        Gary T. Leavens
        329 L3Harris Center
        Computer Science, University of Central Florida
        4000 Central Florida Blvd., Orlando, FL 32816-2362 USA
        http://www.cs.ucf.edu/~leavens phone: +1-407-823-4758
        ***@***.******@***.***>

**Comment by marcel139** [comment]
:-( @davidcok Any help here?

**Comment by marcel139** [comment]
I managed to find a (non-ideal) workaround. ~

1) Add a static method print that encapsulates the invocation to println and use this method rather than println

```
public class CircleOperations {
    public int id(int radius) {
        return radius;
    }
    public static void main(String [] argv) {
        CircleOperations cops = new CircleOperations();
        print(cops.id(79));
    }
    public static void print(int i) {
        System.out.println(i);
    }
}
```

2) Use the option --exclude CircleOperations.print

`./openjml --esc --exclude CircleOperations.print /media/sf_VirtualBox/marcel/ajml/ready/CircleOperations.java`

Of course, this is far from ideal but allows us to print and verify. ;-)

**Comment by marcel139** [comment]
A second (yet non-ideal, but better) workaround:

1) Define a Printer class as follows (attention to the `skipesc` annotation):

```
public class Printer {
    //@ skipesc
    public static void print(String s) {
       System.out.println("Finished");
    }
}

```
2) Use that method in the CircleOperations class

```
public class CircleOperations {

    public int id(int radius) {
        return radius;
    }
    
    public static void main(String [] argv) {
        CircleOperations cops = new CircleOperations();
        Printer.print("Finished");
    }
}
```

3) Use the option --source-path in the command line to make Printer.java visible to the command

`./openjml --esc --source-path /media/sf_VirtualBox/marcel/ajml/ready /media/sf_VirtualBox/marcel/ajml/ready/CircleOperations.java`

**Comment by davidcok** [comment]
Marcel - what is the version of Java on your system (java -version)? If not 17, try running with that.

**Comment by marcel139** [comment]
> Marcel - what is the version of Java on your system (java -version)? If not 17, try running with that.

Hi David,

A quick context. I am not a Linux user, so I have been using a Ubuntu 20.04 distribution running on Oracle VM Virtual Box 7.0. 

For this reason, this virtual machine has only OpenJML "installed". As expected, the command `java -version` returns an error message saying that `java could not be found, but it can be installed...` .

I have been testing openjml ESC for near a month with success. Is the Java distribution really a matter for using --esc?

By the way, here is the content of version-info.txt

```
0.17.0-alpha-15
OpenJML 9d4e5fa12ded902ff846257e26b9a113974ffec8
Specs 45d631e7c9d4caa332e28121b19b603378eb0d02
JMLAnnotations a5dabe3d4af6c4265fdf23c3519ca91ac158faa5
Solvers 1b3073a5b31c07ec95c32c599989bf918b0ff82a
openjml.github.io 0f5cfc83348a6b637baa14b6398139461a6db73f
```

**Comment by marcel139** [comment]
I found out that by adding `pure` to the method specification, everything works fine (see below). 

```
public class CircleOperations {
    //@ pure
    public int id(int radius) {
        return radius;
    }
    public static void main(String [] argv) {
        CircleOperations cops = new CircleOperations();
        System.out.println(cops.id(79));
    }
}

```

I assume that OpenJML has some stuff regarding output streaming that I could not find in the documentation. So, by specifying the framing conditions of the invoked methods we ensure that the method does not mess up the things related to the output.

@davidcok Is there any material that we could read regarding this matter?

**Comment by leavens** [comment]
Hi all,

That is strange, but I can confirm it, and the fact that putting the pure attribute in the declaration fixes the problem, even when println is called twice (see the attached for the different versions of the Java program with JML specifications):

$ uname -a
Linux NEW-SURFACE 5.10.102.1-microsoft-standard-WSL2 #1 SMP Wed Mar 2 00:30:59 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
$ java -version
openjdk version "17.0.6" 2023-01-17
OpenJDK Runtime Environment (build 17.0.6+10-Ubuntu-0ubuntu122.04)
OpenJDK 64-Bit Server VM (build 17.0.6+10-Ubuntu-0ubuntu122.04, mixed mode, sharing)
$ esc -version
openjml 0.17.0-alpha-15
$ esc CircleOperations.java
$ echo $?
0
$ esc CircleOperationsNotPure.java
CircleOperationsNotPure.java:9: verify: The prover cannot establish an assertion (InvariantLeaveCaller: /usr/local/openjml/specs/java/io/PrintStream.jml:42:) in method main: (Caller: CircleOperationsNotPure.main(java.lang.String[]), Callee: java.io.PrintStream.println(int))
        System.out.println(cops.id(79));
                          ^
/usr/local/openjml/specs/java/io/PrintStream.jml:42: verify: Associated declaration: CircleOperationsNotPure.java:9:
    //-RAC@ public invariant outputText != null && \invariant_for(outputText);
                   ^
CircleOperationsNotPure.java:9: verify: The prover cannot establish an assertion (InvariantLeaveCaller: /usr/local/openjml/specs/java/io/PrintStream.jml:35:) in method main: (Caller: CircleOperationsNotPure.main(java.lang.String[]), Callee: java.io.PrintStream.println(int))
        System.out.println(cops.id(79));
                          ^
/usr/local/openjml/specs/java/io/PrintStream.jml:35: verify: Associated declaration: CircleOperationsNotPure.java:9:
    //-RAC@ public static invariant \invariant_for(eol) && eol.length() > 0;
                          ^
4 verification failures
$ esc CircleOperations2Prints.java


        Regards,

        Gary T. Leavens
        329 L3Harris Center
        Computer Science, University of Central Florida
        4000 Central Florida Blvd., Orlando, FL 32816-2362 USA
        http://www.cs.ucf.edu/~leavens phone: +1-407-823-4758
        ***@***.******@***.***>

**Comment by davidcok** [comment]
OK - sorry I've been too tied up with vacation/another deadline to even have read this until now. The problem is this. Without the pure annotation. the id method is `assigns \everything`, which means that any of the internal fields of System.out might be changed, which means that the invariants etc. cannot be reestablished.

I wonder if a general warning on anything without a pure or explicit assigns clause should be issued.

**Comment by marcel139** [comment]
As I suspected. Thanks David.

As a matter of fact, I would consider a good practice to always have pure or explicit assigns on every method/constructor. I would be happy to have such a warning.

**Comment by davidcok** [comment]
Fixed in V21-0.3
